addons/auth_signup/controllers/main.py

```diff
@@ -42,6 +42,12 @@ def web_auth_signup(self, *args, **kw):
         if 'error' not in qcontext and request.httprequest.method == 'POST':
             try:
                 self.do_signup(qcontext)
+
+                # set user to public if they were not signed in by do_signup (mfa enabled)
+                if request.session.uid is None:
+                    public_user = request.env.ref('base.public_user')
+                    request.update_env(user=public_user)
```

Comment:
I am not familiar at all with the sign in / sign up code. Could you explain to me what makes it so enforce the "public user" would lead to the 2FA page but "no user" would lead to an error page ? Thank you :pray:

Reply:
The `do_signup()` method called just above also authenticates the user. When 2FA is enforced, the authentication is partial (the `finalize()` method isn't called). In a clean session, the partial authentication results in the user being set to `None`.

After signing up, an email is sent to the user to notify they created a new account. The email template is rendered using `sudo()`. Unfortunately, during the template rendering, the `sudo()` context is lost in subsequent calls to `ir_http.url_lang()` and `Lang.get_available()`'s context is reverted back to the request.

Because the request context's user is `None`, no model is accessible (no permission). But `Lang.get_available()` needs read access to `website`, which it therefor doesn't have. Hence 403.

My solution, making sure user is set to public gives read access to the `website` among other things and fixes the 403. However, I believe the best way to fix this would be to avoid losing the `sudo()` context.